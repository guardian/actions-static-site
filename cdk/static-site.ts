import {
  GuDistributionBucketParameter,
  GuStack,
  GuStackProps,
} from "@guardian/cdk/lib/constructs/core";
import { App, Duration, SecretValue } from "aws-cdk-lib";
import {
  ApplicationLoadBalancer,
  ApplicationProtocol,
  ListenerAction,
} from "aws-cdk-lib/aws-elasticloadbalancingv2";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { GuVpc, SubnetType } from "@guardian/cdk/lib/constructs/ec2";
import { GuCname } from "@guardian/cdk/lib/constructs/dns/";
import { GuCertificate } from "@guardian/cdk/lib/constructs/acm";
import { Code, Function, LayerVersion, Runtime } from "aws-cdk-lib/aws-lambda";
import { LambdaTarget } from "aws-cdk-lib/aws-elasticloadbalancingv2-targets";
import { StringParameter } from "aws-cdk-lib/aws-ssm";
import { Port } from "aws-cdk-lib/aws-ec2";

interface StaticSiteProps extends GuStackProps {
  app: string;
  domainName: string;
  auth?: boolean;
  layerVersionName?: string; // Intended purely to stabilise for snapshot testing.
}

// It is surprisingly tricky in AWS to setup a static site, with a custom domain
// and Google authentication. The solution chosen here is as follows:
//
// ALB -> Lambda -> Layer
//
// where Google authentication is performed at the ALB level, and the site is
// served from the Lambda as a layer. The Lambda is a simple Go proxy to the
// layer.
//
// Users of this action should NOT assume that this architecture won't change
// going forward though.
//
// **Note:** layers have a size limit of 250mb. So your site cannot exceed this
// in size.
export class StaticSite extends GuStack {
  constructor(scope: App, id: string, props: StaticSiteProps) {
    super(scope, id, props);

    const bucket = Bucket.fromBucketName(
      this,
      `${id}-bucket`,
      GuDistributionBucketParameter.getInstance(this).valueAsString
    );

    const s3Prefix = `${this.stack}/${this.stage}/${props.app}`;

    const siteLayer = new LayerVersion(this, "site-layer", {
      code: Code.fromBucket(bucket, `${s3Prefix}/site.zip`),
      description: "layer for static site",
      layerVersionName:
        props.layerVersionName ??
        `site-layer-${Math.floor(new Date().getTime() / 1000)}`,
    });

    const lambda = new Function(this, "lambda", {
      runtime: Runtime.GO_1_X,
      handler: "main",
      code: Code.fromBucket(bucket, `${s3Prefix}/lambda.zip`),
      layers: [siteLayer],
    });

    const vpc = GuVpc.fromIdParameter(this, "vpc-id");
    const publicSubnets = GuVpc.subnetsFromParameter(this, {
      type: SubnetType.PUBLIC,
      app: props.app,
    });

    const alb = new ApplicationLoadBalancer(this, "alb", {
      vpc: vpc,
      vpcSubnets: { subnets: publicSubnets },
      internetFacing: true,
    });

    // Ensure ALB can reach ldp userInfo endpont - see
    // https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-troubleshooting.html#http-500-issues.
    alb.connections.allowToAnyIpv4(Port.tcp(443));

    new GuCname(this, "cname", {
      app: props.app,
      domainName: props.domainName,
      ttl: Duration.minutes(1),
      resourceRecord: alb.loadBalancerDnsName,
    });

    const cert = new GuCertificate(this, {
      app: props.app,
      domainName: props.domainName,
    });

    const listener = alb.addListener("listener", {
      protocol: ApplicationProtocol.HTTPS,
      port: 443,
      certificates: [cert],
    });

    const targetGroup = listener.addTargets("target", {
      targets: [new LambdaTarget(lambda)],
    });
    targetGroup.setAttribute("lambda.multi_value_headers.enabled", "true"); // See: https://github.com/akrylysov/algnhsa/pull/20/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R81.

    if (props.auth) {
      const ssmPrefix = `${this.stage}/${this.stack}/${props.app}`;
      const clientId = StringParameter.fromStringParameterAttributes(
        this,
        "clientID",
        { parameterName: `/${ssmPrefix}/googleClientID` }
      ).stringValue;

      // Secure SSM Parameters here aren't possible unfortunately
      // (Cloudformation only provides narrow support for them at the time of
      // writing). See:
      // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references.html#template-parameters-dynamic-patterns-resources.
      const clientSecret = SecretValue.secretsManager(
        `${ssmPrefix}/googleClientSecret`
      );

      const authAction = ListenerAction.authenticateOidc({
        next: ListenerAction.forward([targetGroup]),

        clientId: clientId,
        clientSecret: clientSecret,

        scope: "openid email",

        // See the `hd` section of https://developers.google.com/identity/protocols/oauth2/openid-connect#authenticationuriparameters.
        authenticationRequestExtraParams: { hd: "guardian.co.uk" },

        authorizationEndpoint: "https://accounts.google.com/o/oauth2/v2/auth",
        issuer: "https://accounts.google.com",
        tokenEndpoint: "https://oauth2.googleapis.com/token",
        userInfoEndpoint: "https://openidconnect.googleapis.com/v1/userinfo",
      });

      listener.addAction("auth", { action: authAction });
    }
  }
}
