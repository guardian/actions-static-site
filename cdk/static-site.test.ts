import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { StaticSite } from "./static-site";

describe("static site", () => {
  it("should match snapshot", () => {
    const cdkApp = new App();
    const cdkStack = new StaticSite(cdkApp, "static-site", {
      app: 'app',
      stack: 'stack',
      stage: "PROD",
      domainName: "example.devx.gutools.co.uk",
      layerVersionName: "layer",
      auth: true,
    });

    const got = Template.fromStack(cdkStack).toJSON();
    expect(got).toMatchSnapshot()
  })
})
