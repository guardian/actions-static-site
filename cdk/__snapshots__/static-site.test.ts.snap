// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`static site should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "VpcId": Object {
      "Default": "/account/vpc/primary/id",
      "Description": "Virtual Private Cloud to run EC2 instances within. Should NOT be the account default VPC.",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
    },
    "appPublicSubnets": Object {
      "Default": "/account/vpc/primary/subnets/public",
      "Description": "A list of public subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
  },
  "Resources": Object {
    "CertificateAppCEE540C3": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "DomainName": "example.devx.gutools.co.uk",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "app",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "46.0.1",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/actions-static-site",
          },
          Object {
            "Key": "Stack",
            "Value": "stack",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
      "UpdateReplacePolicy": "Retain",
    },
    "alb8A8B13C2": Object {
      "Properties": Object {
        "LoadBalancerAttributes": Array [
          Object {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "albSecurityGroup49866104",
              "GroupId",
            ],
          },
        ],
        "Subnets": Object {
          "Ref": "appPublicSubnets",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "46.0.1",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/actions-static-site",
          },
          Object {
            "Key": "Stack",
            "Value": "stack",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "albSecurityGroup49866104": Object {
      "Properties": Object {
        "GroupDescription": "Automatically created Security Group for ELB staticsitealb4CA23368",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "46.0.1",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/actions-static-site",
          },
          Object {
            "Key": "Stack",
            "Value": "stack",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "VpcId": Object {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "alblistenerC508FBF5": Object {
      "Properties": Object {
        "Certificates": Array [
          Object {
            "CertificateArn": Object {
              "Ref": "CertificateAppCEE540C3",
            },
          },
        ],
        "DefaultActions": Array [
          Object {
            "TargetGroupArn": Object {
              "Ref": "alblistenertargetGroup2F1EE73B",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": Object {
          "Ref": "alb8A8B13C2",
        },
        "Port": 443,
        "Protocol": "HTTPS",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "alblistenertargetGroup2F1EE73B": Object {
      "DependsOn": Array [
        "lambdaInvoke2UTWxhlfyqbT5FTn5jvgbLgjFfJwzswGk55DU1HY67C4BB0E",
      ],
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "46.0.1",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/actions-static-site",
          },
          Object {
            "Key": "Stack",
            "Value": "stack",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "TargetType": "lambda",
        "Targets": Array [
          Object {
            "Id": Object {
              "Fn::GetAtt": Array [
                "lambda8B5974B5",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "cname": Object {
      "Properties": Object {
        "Name": "example.devx.gutools.co.uk",
        "RecordType": "CNAME",
        "ResourceRecords": Array [
          Object {
            "Fn::GetAtt": Array [
              "alb8A8B13C2",
              "DNSName",
            ],
          },
        ],
        "Stage": "PROD",
        "TTL": 60,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
    "lambda8B5974B5": Object {
      "DependsOn": Array [
        "lambdaServiceRole494E4CA6",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "stack/PROD/app/lambda.zip",
        },
        "Handler": "main",
        "Layers": Array [
          Object {
            "Ref": "sitelayerD1248BA7",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "lambdaServiceRole494E4CA6",
            "Arn",
          ],
        },
        "Runtime": "go1.x",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "46.0.1",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/actions-static-site",
          },
          Object {
            "Key": "Stack",
            "Value": "stack",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaInvoke2UTWxhlfyqbT5FTn5jvgbLgjFfJwzswGk55DU1HY67C4BB0E": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "lambda8B5974B5",
            "Arn",
          ],
        },
        "Principal": "elasticloadbalancing.amazonaws.com",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "lambdaServiceRole494E4CA6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "46.0.1",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/actions-static-site",
          },
          Object {
            "Key": "Stack",
            "Value": "stack",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "sitelayerD1248BA7": Object {
      "Properties": Object {
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "stack/PROD/app/site.zip",
        },
        "Description": "layer for static site",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
  },
}
`;
